File: rb/test/symbol_entity_test.rb
# Symbol entity test

require "minitest/autorun"
require "json"
require_relative "../ExchangeRates_sdk"
require_relative "runner"

class SymbolEntityTest < Minitest::Test
  def test_create_instance
    testsdk = ExchangeRatesSDK.test(nil, nil)
    ent = testsdk.Symbol(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = symbol_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "symbol." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set EXCHANGERATES_TEST_SYMBOL_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    symbol_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.symbol")))
    symbol_ref01_data = nil
    if symbol_ref01_data_raw.length > 0
      symbol_ref01_data = Helpers.to_map(symbol_ref01_data_raw[0][1])
    end

    # LOAD
    symbol_ref01_ent = client.Symbol(nil)
    symbol_ref01_match_dt0 = {}
    symbol_ref01_data_dt0_loaded = symbol_ref01_ent.load(symbol_ref01_match_dt0, nil)
    assert !symbol_ref01_data_dt0_loaded.nil?

  end
end

def symbol_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "symbol", "SymbolTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = ExchangeRatesSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["symbol01", "symbol02", "symbol03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["EXCHANGERATES_TEST_SYMBOL_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "EXCHANGERATES_TEST_SYMBOL_ENTID" => idmap,
    "EXCHANGERATES_TEST_LIVE" => "FALSE",
    "EXCHANGERATES_TEST_EXPLAIN" => "FALSE",
    "EXCHANGERATES_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["EXCHANGERATES_TEST_SYMBOL_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["EXCHANGERATES_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["EXCHANGERATES_APIKEY"],
      },
      extra || {},
    ])
    client = ExchangeRatesSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["EXCHANGERATES_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["EXCHANGERATES_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: ts/test/entity/symbol/SymbolEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { ExchangeRatesSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('SymbolEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when EXCHANGE_RATES_TEST_LIVE=TRUE.
  afterEach(liveDelay('EXCHANGE_RATES_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = ExchangeRatesSDK.test()
    const ent = testsdk.Symbol()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.EXCHANGE_RATES_TEST_LIVE
    for (const op of ['load']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'symbol.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"country","req":true,"short":"Country or region","type":"`$STRING`","index$":0},{"active":true,"name":"name","req":true,"short":"Full name of the currency","type":"`$STRING`","index$":1},{"active":true,"name":"symbol","req":true,"short":"Currency symbol","type":"`$STRING`","index$":2}],"name":"symbol","op":{"load":{"input":"data","name":"load","points":[{"active":true,"args":{},"contract":{"id":"GET /symbols","json":"{\"operationId\":\"getSupportedCurrencies\",\"parameters\":[],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"example\":{\"base\":\"AUD\",\"count\":21,\"note\":\"All rates are quoted as AUD per unit of foreign currency from the Reserve Bank of Australia\",\"success\":true,\"symbols\":{\"EUR\":{\"country\":\"European Union\",\"name\":\"Euro\",\"symbol\":\"€\"},\"USD\":{\"country\":\"United States\",\"name\":\"US Dollar\",\"symbol\":\"$\"}}},\"schema\":{\"properties\":{\"base\":{\"example\":\"AUD\",\"type\":\"string\"},\"count\":{\"description\":\"Number of supported currencies\",\"type\":\"integer\"},\"note\":{\"type\":\"string\"},\"success\":{\"example\":true,\"type\":\"boolean\"},\"symbols\":{\"additionalProperties\":{\"properties\":{\"country\":{\"description\":\"Country or region\",\"type\":\"string\"},\"name\":{\"description\":\"Full name of the currency\",\"type\":\"string\"},\"symbol\":{\"description\":\"Currency symbol\",\"type\":\"string\"}},\"required\":[\"name\",\"symbol\",\"country\"],\"type\":\"object\"},\"type\":\"object\"}},\"required\":[\"success\",\"symbols\",\"count\",\"base\",\"note\"],\"type\":\"object\"}}},\"description\":\"List of supported currencies\"}},\"securitySchemes\":{\"bearerAuth\":{\"bearerFormat\":\"API Key\",\"description\":\"API key authentication using Bearer token\",\"scheme\":\"bearer\",\"type\":\"http\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/symbols","segments":[{"lit":"symbols"}],"select":{},"transform":{"req":"`reqdata`","res":"`body.symbols`"},"index$":0}],"key$":"load"}},"relations":{"ancestors":[]},"key$":"symbol","name__orig":"symbol","Name":"Symbol","name_":"symbol","name-":"symbol","NAME":"SYMBOL","index$":6}, {"active":true,"entity":"symbol","key$":"BasicSymbolFlow","kind":"basic","name":"BasicSymbolFlow","param":{},"step":[{"active":true,"data":{},"input":{"ref":"symbol_ref01","srcdatavar":"symbol_ref01_data","suffix":"_dt0"},"match":{},"op":"load","spec":[],"valid":[{"apply":"TextFieldMark","def":{"mark":"Mark01-symbol_ref01"}}],"index$":0}]}, 'Symbol')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let symbol_ref01_data = Object.values(setup.data.existing.symbol)[0] as any

    // LOAD
    const symbol_ref01_ent = client.Symbol()
    const symbol_ref01_match_dt0: any = {}
    const symbol_ref01_data_dt0 = (await symbol_ref01_ent.load(symbol_ref01_match_dt0)).data()
    assert(null != symbol_ref01_data_dt0)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/symbol/SymbolTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = ExchangeRatesSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['symbol01','symbol02','symbol03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'EXCHANGE_RATES_TEST_SYMBOL_ENTID': idmap,
    'EXCHANGE_RATES_TEST_LIVE': 'FALSE',
    'EXCHANGE_RATES_TEST_EXPLAIN': 'FALSE',
    'EXCHANGE_RATES_APIKEY': '',
  })

  idmap = env['EXCHANGE_RATES_TEST_SYMBOL_ENTID']

  const live = 'TRUE' === env.EXCHANGE_RATES_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['EXCHANGE_RATES_TEST_SYMBOL_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new ExchangeRatesSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
        apikey: env.EXCHANGE_RATES_APIKEY,
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.EXCHANGE_RATES_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  
